apis/nucleus/src/components/listbox/assets/get-list-sizes/column-mode.js

```diff
@@ -0,0 +1,40 @@
+import { ITEM_MIN_WIDTH } from './constants';
+
+export default function calculateColumnMode({
+  maxVisibleRows,
+  itemHeight,
+  listCount,
+  listHeight,
+  columnAutoWidth,
+  containerWidth,
+}) {
+  let rowCount;
+  const maxRows = maxVisibleRows?.maxRows || 3;
+  const autoRowCount = Math.floor(listHeight / itemHeight);
+  if (maxVisibleRows.auto !== false) {
+    rowCount = autoRowCount;
+  } else {
+    rowCount = Math.min(listCount, maxRows);
+    const itemsCanFit = itemHeight * rowCount <= listHeight;
+    if (!itemsCanFit) {
+      // If items cannot fit inside height, fall back to auto mode.
+      return calculateColumnMode({
+        maxVisibleRows: { ...maxVisibleRows, auto: true },
+        itemHeight,
+        listCount,
+        listHeight,
+        columnAutoWidth,
+        containerWidth,
+      });
+    }
```

Comment:
I think you would get the same result by limiting the rowCount to not be larger the autoRowCount
```suggestion
    rowCount = Math.min(listCount, maxRows, autoRowCount);
```

Reply:
Yes, that's essentially doing the same thing but shorter 👍 
I will change in the row calculation too, so we can avoid the recursive call.